toolchain/check/impl_validation.cpp

```diff
@@ -0,0 +1,456 @@
+                                    return impl_and_interface.impl_id;
+                                  });
+
+  // All `impl`s we look at here have the same `InterfaceId` (though different
+  // `SpecificInterface`s in their specifics. So we can grab the ImportIRId for
+  // the interface a single time up front.
+  auto interface_decl_id =
+      context.interfaces()
+          .Get(impls_and_interface[0].interface.interface_id)
+          .first_owning_decl_id;
+  auto interface_ir_id = GetIRId(context, interface_decl_id);
+
+  // TODO: We should revisit this and look for a way to do these checks in less
+  // than quadratic time. From @zygoloid: Possibly by converting the set of
+  // impls into a decision tree.
+  for (auto [index_a, impl_a_id] : llvm::enumerate(impl_ids)) {
+    auto impl_a = GetImplInfo(context, impl_a_id);
+
+    if (impl_a.is_final && impl_a.is_local) {
+      // =======================================================================
+      /// Rules for an individual final impl and its internal values.
+      // =======================================================================
+      if (DiagnoseFinalImplNotInSameFileAsRootSelfTypeOrInterface(
+              context, impl_a, interface_ir_id)) {
+        continue;
+      }
+    }
+
+    for (auto impl_b_id : llvm::drop_begin(impl_ids, index_a + 1)) {
+      auto impl_b = GetImplInfo(context, impl_b_id);
+
+      if (!impl_a.is_final && !impl_b.is_final) {
+        // =====================================================================
+        // Rules between two non-final impls.
+        //
+        // It's possible to write invalid non-final impls in two files that do
+        // not know about each other, and they are only invalid when they are
+        // both imported and visible together. So we need to compare non-local
+        // impls with each other here.
+        // =====================================================================
+        if (DiagnoseNonFinalImplsWithSameTypeStructure(context, impl_a,
+                                                       impl_b)) {
+          continue;
+        }
+      } else if (!impl_a.is_final || !impl_b.is_final) {
+        // No rules that involve a final impl need to be enforced on two
+        // imported impls, so we only diagnose if at least one of the two impls
+        // being compared is local.
+        if (!impl_a.is_local && !impl_b.is_local) {
+          continue;
+        }
+
+        // =====================================================================
+        // Rules between final impl and non-final impl.
+        // =====================================================================
+        if (DiagnoseUnmatchableNonFinalImplWithFinalImpl(
+                context, impl_a_id, impl_a, impl_b_id, impl_b)) {
+          continue;
+        }
+      } else if (impl_a.type_structure.IsCompatibleWith(
+                     impl_b.type_structure)) {
```

Comment:
Should not have the same test here and in impl_lookup. One of them is wrong. Looking at the implementation it looks like a symmetric test, so my guess is this is correct and impl_lookup is wrong.

In impl lookup, a `C as I` query can match a `? as I` impl, but a `? as I` query can't match a `C as I` impl.